From ae14a4bd581dc4acaaf31d79517c6bc38ece8992 Mon Sep 17 00:00:00 2001
From: Chen Baozi <chenbaozi@phytium.com.cn>
Date: Mon, 17 Jun 2024 19:33:09 +0800
Subject: [PATCH 115/150] net: macb: Initialize all descriptors of RX rings

This patch resolves the issue by ensuring that each descriptor is
initialized drugint the RX ring setup, preventing any uninitialized
descriptors and therby avoiding potential DMA conflicts under high
traffic conditions.

Signed-off-by: Song Wenting <songwenting@phytium.com.cn>
Signed-off-by: Li Wencheng <liwencheng@phytium.com.cn>
Signed-off-by: Wang Yinfeng <wangyinfeng@phytium.com.cn>
Signed-off-by: Chen Baozi <chenbaozi@phytium.com.cn>
Change-Id: I00cbd9a82fdc2a571ca57a9eaa99ccf5c7eacb52
Signed-off-by: Andrew Powers-Holmes <aholmes@omnom.net>
---
 drivers/net/ethernet/cadence/macb_main.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/net/ethernet/cadence/macb_main.c b/drivers/net/ethernet/cadence/macb_main.c
index f55de33a9fab..a06b996c0116 100644
--- a/drivers/net/ethernet/cadence/macb_main.c
+++ b/drivers/net/ethernet/cadence/macb_main.c
@@ -2839,6 +2839,16 @@ static void gem_init_rings(struct macb *bp)
 		queue->tx_head = 0;
 		queue->tx_tail = 0;
 
+		for (i = 0; i < bp->rx_ring_size; i++) {
+			desc = macb_rx_desc(queue, i);
+			desc->ctrl = 0;
+			/* make sure ctrl is cleared first,
+			 * and bit RX_USED is set to avoid a race.
+			 */
+			dma_wmb();
+			desc->addr |= MACB_BIT(RX_USED);
+		}
+
 		queue->rx_tail = 0;
 		queue->rx_prepared_head = 0;
 
-- 
2.47.0

